subprojects/file-watching/src/main/java/org/gradle/internal/watch/vfs/impl/DefaultWatchableFileSystemRegistry.java

```diff
@@ -0,0 +1,151 @@
+ * limitations under the License.
+ */
+
+package org.gradle.internal.watch.vfs.impl;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+import net.rubygrapefruit.platform.Native;
+import net.rubygrapefruit.platform.file.FileSystemInfo;
+import net.rubygrapefruit.platform.file.FileSystems;
+import org.gradle.internal.watch.WatchingNotSupportedException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.util.Comparator;
+import java.util.List;
+
+public class DefaultWatchableFileSystemRegistry implements WatchableFileSystemRegistry {
+    private static final Logger LOGGER = LoggerFactory.getLogger(DefaultWatchableFileSystemRegistry.class);
+
+    private static final ImmutableSet<String> SUPPORTED_FILE_SYSTEM_TYPES = ImmutableSet.of(
+        // APFS on macOS
+        "apfs",
+        // HFS and HFS+ on macOS
+        "hfs",
+        "ext3",
+        "ext4",
+        "btrfs",
+        // NTFS on macOS
+        "ntfs",
+        // NTFS on Windows
+        "NTFS",
+        // FAT32 on macOS
+        "msdos",
+        // FAT32 on Windows
+        "FAT32",
+        // exFAT on macOS
+        "exfat",
+        // exFAT on Windows
+        "exFAT",
+        // VirtualBox FS
+        "vboxsf"
+    );
+
+    private final ImmutableList<FileSystemSupport> fileSystemRoots;
+
+    public static WatchableFileSystemRegistry create() {
+        return new DefaultWatchableFileSystemRegistry(Native.get(FileSystems.class).getFileSystems());
+    }
+
+    @VisibleForTesting
+    DefaultWatchableFileSystemRegistry(List<FileSystemInfo> fileSystems) {
+        ImmutableList.Builder<FileSystemSupport> builder = ImmutableList.<FileSystemSupport>builder();
+        fileSystems.stream()
+            // Sort by longest path first so we always match most the specific location in case locations are nested
+            .sorted(Comparator.comparingInt(fileSystem -> -fileSystem.getMountPoint().getAbsolutePath().length()))
+            .map(FileSystemSupport::new)
+            .forEach(support -> {
+                LOGGER.info("Detected {} {}: {} from {} (remote: {}, case-sensitive: {}, case-preserving: {})",
+                    support.isSupported() ? "supported" : "unsupported",
+                    support.getFileSystem().getFileSystemType(),
+                    support.getPrefix(),
+                    support.getFileSystem().getDeviceName(),
+                    support.getFileSystem().isRemote(),
+                    support.getFileSystem().isCaseSensitive(),
+                    support.getFileSystem().isCasePreserving());
+                builder.add(support);
+            });
+        this.fileSystemRoots = builder.build();
+    }
+
+    @Override
+    public void ensureWatchingSupported(File path) {
+        String prefix = toAbsolutePathPrefix(path);
+        for (FileSystemSupport support : fileSystemRoots) {
```

Comment:
How about we check that all file systems under need to support watching? That way we don't watch if any of the file systems doesn't support watching.

Reply:
There are always file systems that don't support watching on *NIX's, like `/dev` and `/proc` etc.

Reply:
I mean that all filesystem under the given path (aka project root) need to support watching. Not all file systems at the root of the file system (aka '/`).